packages/core/src/validate/run.ts

```diff
@@ -206,27 +225,159 @@ function* getConstructorErrors(contractDef: ContractDefinition, decodeSrc: SrcDe
+  if (visitedNodeIds.has(contractDef.id)) {
+    return;
+  } else {
+    visitedNodeIds.add(contractDef.id);
+  }
+
+  yield* getFunctionOpcodeErrors(contractDef, deref, decodeSrc, opcode, skipInternal, visitedNodeIds);
+  yield* getInheritedContractOpcodeErrors(contractDef, deref, decodeSrc, opcode, visitedNodeIds);
+}
+
+function* getFunctionOpcodeErrors(
+  contractOrFunctionDef: ContractDefinition | FunctionDefinition,
+  deref: ASTDereferencer,
+  decodeSrc: SrcDecoder,
+  opcode: OpcodePattern,
+  skipInternal: boolean,
+  visitedNodeIds: Set<number>,
+): Generator<ValidationErrorOpcode> {
+  const parentNode = getParentNode(deref, contractOrFunctionDef);
+  if (parentNode === undefined || !skipCheck(opcode.kind, parentNode)) {
+    yield* getDirectFunctionOpcodeErrors(contractOrFunctionDef, decodeSrc, opcode, skipInternal);
+  }
+  if (parentNode === undefined || !skipCheckReachable(opcode.kind, parentNode)) {
+    yield* getReferencedFunctionOpcodeErrors(
+      contractOrFunctionDef,
+      deref,
+      decodeSrc,
+      opcode,
+      skipInternal,
+      visitedNodeIds,
+    );
+  }
+}
+
+function* getDirectFunctionOpcodeErrors(
+  contractOrFunctionDef: ContractDefinition | FunctionDefinition,
+  decodeSrc: SrcDecoder,
+  opcode: OpcodePattern,
+  skipInternal: boolean,
+) {
+  for (const fnCall of findAll(
+    'FunctionCall',
+    contractOrFunctionDef,
+    node => skipCheck(opcode.kind, node) || skipInternalFunctions(skipInternal, node),
+  )) {
     const fn = fnCall.expression;
-    if (fn.typeDescriptions.typeIdentifier?.match(/^t_function_baredelegatecall_/)) {
+    if (fn.typeDescriptions.typeIdentifier?.match(opcode.pattern)) {
       yield {
-        kind: 'delegatecall',
+        kind: opcode.kind,
         src: decodeSrc(fnCall),
       };
     }
   }
-  for (const fnCall of findAll('FunctionCall', contractDef, node => skipCheck('selfdestruct', node))) {
+}
+
+function* getReferencedFunctionOpcodeErrors(
+  contractOrFunctionDef: ContractDefinition | FunctionDefinition,
+  deref: ASTDereferencer,
+  decodeSrc: SrcDecoder,
+  opcode: OpcodePattern,
+  skipInternal: boolean,
+  visitedNodeIds: Set<number>,
+) {
+  for (const fnCall of findAll(
+    'FunctionCall',
+    contractOrFunctionDef,
+    node => skipCheckReachable(opcode.kind, node) || skipInternalFunctions(skipInternal, node),
+  )) {
     const fn = fnCall.expression;
-    if (fn.typeDescriptions.typeIdentifier?.match(/^t_function_selfdestruct_/)) {
-      yield {
-        kind: 'selfdestruct',
-        src: decodeSrc(fnCall),
-      };
+    const fnReference = (fn as any).referencedDeclaration;
+    if (fnReference !== undefined && fnReference > 0) {
+      try {
+        const referencedNode = deref(['FunctionDefinition'], fnReference);
+        if (!visitedNodeIds.has(referencedNode.id)) {
+          visitedNodeIds.add(referencedNode.id);
+          yield* getFunctionOpcodeErrors(referencedNode, deref, decodeSrc, opcode, false, visitedNodeIds);
```

Comment:
Hm... I'm slightly concerned that we would visit a node with `skipInternal=true` _before_ we visit it with `skipInternal=false`, thus skipping over it the second time and potentially missing errors.

My intuition is that this is not happening but only becasue we run `getFunctionOpcodeErrors` before we run `getInheritedContractOpcodeErrors`. This seems fragile. I would appreciate if the correctness relied on a more explicit mechanism. For example, instead of a `Set` for visited nodes, we could have a `Map<number, 'internal' | 'all'>` where we explicitly mark visited and visibility.

Reply:
Addressed in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/702/commits/6f47413d1db5a208abe2c75e926814338ed6a40e